proposals/dictionary-expressions.md

```diff
@@ -33,13 +33,20 @@ An inclusive solution is needed for C#. It should meet the vast majority of case
 The following grammar productions are added:
 
 ```diff
++ collection_expression
+  : '[' ']'
+- | '[' collection_element ( ',' collection_element )* ']'  
++ | '[' 'comparer': expression ']'
+  | '[' ('comparer': expression , )? collection_element ( ',' collection_element )* ']'
+  ;
```

Comment:
this is new.  I'm taking an opinionated stance that we just support `comparers`.  We haven't supported any further customization around normal collection expressions, and feedback has been minimal.  I do think comparers are relevant enough for dictionaries (*especially* the preponderance of case insensitive string dictionaries) to warrant syntax.

This syntax isn't great, given that it overlaps `k:v` elements.  However, in practice i expect any collisions to be rare.  And, if they occur, users can write: `@comparer: ...`.  Finally, the IDE will classify `comparer` as a keyword, making it clear that that's what is being provided.

Reply:
Not spaceship time? /s
`[<=>: StringComparer.OrdinalIgnoreCase, "a": a, "b": b]`

Reply:
I'd like to consider whether we need this versus telling people to write `TheDictType.Create([a: b, c: d], StringComparer.OrdinalIgnoreCase)`. The BCL could just provide an overload or optional parameter for the comparer, and the performance of this construction path would be no worse than any general CollectionBuilder collection.

Reply:
So compare these for convenience.

- `[comparer: StringComparer.OrdinalIgnoreCase, a: b, c: d]`
- `TheDictType.Create([a: b, c: d], StringComparer.OrdinalIgnoreCase)`
- `new([a: b, c: d], StringComparer.OrdinalIgnoreCase)`

The latter case would broaden the meaning of `new(` to say it can call constructors and CollectionBuilder Create methods.

Reply:
i don't think that helps, especially needing a solution for every type.  it also greatly increases teh verbosity.

My thinking on why i want to support this special guy is simply that how hashing/equality working is fundamental to what a dictionary is.  This is unlike other arguments which so far have been more about optimizing performance.

Reply:
I think our posts crossed, but I want to push back on "greatly increases the verbosity." See https://github.com/dotnet/csharplang/pull/8568/files#r1835510207

That is a minor increase in verbosity IMO. And it's lowering the concept count for understanding the syntax, so it pays us back.

> especially needing a solution for every type
Only those types which don't have a `set;` in their indexer, but this is true.

The compiler could address this by resolving `new(` to a constructor with the same parameter set minus the elements, and then sending the elements to the indexer.

Reply:
Cyrus brought up a great point offline, which is that we'd like to support targeting interface dictionary types with comparers. I agree that this is an important use case, and it definitely rules out `new(` and `Create(` approaches.